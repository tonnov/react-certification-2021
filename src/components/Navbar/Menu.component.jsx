import React, { useState } from 'react';
import MenuItem from '@material-ui/core/MenuItem';
import { withStyles } from '@material-ui/core/styles';
import Menu from '@material-ui/core/Menu';
import PersonIcon from '@material-ui/icons/Person';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Login from '../Login';
import { useAuth } from '../../providers/Auth';

const MenuLogin = ({ anchor, toggleMenu, userName }) => {
  const StyledMenu = withStyles({
    paper: {
      border: '1px solid #d3d4d5',
    },
  })((props) => (
    <Menu
      elevation={0}
      getContentAnchorEl={null}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'center',
      }}
      transformOrigin={{
        vertical: 'top',
        horizontal: 'center',
      }}
      {...props}
    />
  ));

  const [loginForm, setLoginForm] = useState(false);

  const showLogin = () => {
    toggleMenu(null);
    setLoginForm(!loginForm);
  };

  const { authenticated, logout } = useAuth();

  const handleClose = () => {
    toggleMenu(null);
  };

  const handleLogout = () => {
    logout();
    toggleMenu(null);
  };

  return (
    <>
      <Login show={loginForm} toggleShow={showLogin} />
      <StyledMenu
        anchorEl={anchor}
        keepMounted
        open={Boolean(anchor)}
        onClose={handleClose}
      >
        {authenticated ? (
          <span>
            <ListItemIcon aria-label="User Active">
              <PersonIcon />
              <ListItemText fontSize="small" primary={userName} />
            </ListItemIcon>
            <MenuItem aria-label="Log Out" onClick={handleLogout}>
              Log Out
            </MenuItem>
          </span>
        ) : (
          <MenuItem aria-label="Log In" onClick={showLogin}>
            Log In
          </MenuItem>
        )}
      </StyledMenu>
    </>
  );
};

export default MenuLogin;
