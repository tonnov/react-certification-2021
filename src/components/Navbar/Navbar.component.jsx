import React, { useState, useEffect } from 'react';
import { IconButton, FormControlLabel } from '@material-ui/core';
import Switch from '@material-ui/core/Switch';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import { useHistory } from 'react-router-dom';
import {
  GlobalStyle,
  Nav,
  NavLeft,
  NavRight,
  NavSwitch,
  UserAvatar,
  SearchContainer,
  SearchInput,
  IconContainer,
  AuthAvatar,
} from './Navbar.styled';
import FullSideBar from './Sidebar.component';
import { useDebounce } from '../../utils/hooks/useDebounce';
import { useGlobal } from '../../providers/Global';
import MenuLogin from './Menu.component';
import { useAuth } from '../../providers/Auth';

function Navbar() {
  const [sidebar, setSidebar] = useState(false);
  const [anchorEl, setAnchorEl] = useState(null);

  const { authenticated } = useAuth();
  const { state, dispatch } = useGlobal();
  const { query, darkTheme, sessionData } = state;

  const [localQuery, setLocalQuery] = useState(query);

  const history = useHistory();

  const debounceValue = useDebounce(localQuery, 500);

  useEffect(() => {
    if (!debounceValue) return;

    dispatch({ type: 'update_search_query', payload: debounceValue });
    history.push('/');
  }, [debounceValue, history, dispatch]);

  const showSidebar = () => setSidebar(!sidebar);

  const showMenu = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const hideMenu = () => setAnchorEl(null);

  const changeLocalQuery = (e) => {
    setLocalQuery(e.target.value);
  };

  const handleChangeTheme = () => {
    dispatch({ type: 'toggle_theme' });
  };

  const menuItems = [{ route: '/', name: 'Home' }];

  if (authenticated) {
    menuItems.push({ route: '/favorites', name: 'Favorites' });
  }

  return (
    <>
      <GlobalStyle dark={darkTheme} />
      <Nav dark={darkTheme} aria-label="NavBar">
        <MenuLogin anchor={anchorEl} toggleMenu={hideMenu} userName={sessionData?.name} />
        <NavLeft>
          <IconButton color="inherit" aria-label="open drawer" onClick={showSidebar}>
            <MenuIcon />
          </IconButton>
          <FullSideBar
            dark={darkTheme}
            updateState={showSidebar}
            activeClass={sidebar ? 'sidebar-active' : null}
            menuItems={menuItems}
            updateTheme={handleChangeTheme}
          />
          <SearchContainer>
            <IconContainer>
              <SearchIcon />
            </IconContainer>
            <SearchInput
              placeholder="Search..."
              value={localQuery}
              onChange={changeLocalQuery}
              aria-label="search"
            />
          </SearchContainer>
        </NavLeft>
        <NavRight>
          <NavSwitch>
            <FormControlLabel
              control={
                <Switch
                  checked={darkTheme}
                  name="toggleTheme"
                  color="primary"
                  onChange={handleChangeTheme}
                />
              }
              label="Dark mode"
            />
          </NavSwitch>
          <IconButton color="inherit" aria-label="user" onClick={showMenu}>
            {authenticated ? (
              <AuthAvatar url={sessionData?.avatarUrl} />
            ) : (
              <UserAvatar dark={darkTheme} />
            )}
          </IconButton>
        </NavRight>
      </Nav>
    </>
  );
}

export default Navbar;
