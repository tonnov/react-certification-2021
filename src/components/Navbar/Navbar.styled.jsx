import React from 'react';
import styled, { createGlobalStyle } from 'styled-components';
import PersonIcon from '@material-ui/icons/Person';

export const GlobalStyle = createGlobalStyle`
  body {
    background-color: ${(props) => (props.dark ? '#3b3b3b' : '#fff')};
    color: ${(props) => (props.dark ? '#f2f2f2' : '#3d3d3d')};
  }
`;

export const Nav = styled.nav`
  background-color: ${(props) => (props.dark ? '#556CD6' : '#1c5476')};
  color: #fafafa;
  height: 64px;
  width: 100%;
  display: flex;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.2);
  @media screen and (max-width: 600px) {
    height: 56px;
  }
`;

export const NavLeft = styled.div`
  flex: 50%;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  margin-left: 15px;
  @media screen and (max-width: 600px) {
    flex: 95%;
    margin-left: 10px;
  }
`;

export const NavRight = styled.div`
  flex: 50%;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  margin-right: 25px;
  @media screen and (max-width: 600px) {
    flex: 5%;
    margin-right: 0;
  }
`;

export const NavSwitch = styled.span`
  @media screen and (max-width: 600px) {
    display: none;
  }
`;

export const SearchContainer = styled.div`
  position: relative;
  width: 240px;
  height: 60%;
  margin-left: 10px;
  background-color: rgba(255, 255, 255, 0.15);
  border-radius: 4px;
  display: flex;
  align-items: center;
  padding-left: 5px;

  :hover {
    transition: all 0.2s ease;
    background-color: rgba(255, 255, 255, 0.3);
  }
  @media screen and (max-width: 600px) {
    height: 65%;
    width: 85%;
  }
`;

export const SearchInput = styled.input`
  flex: 90%;
  height: 100%;
  background-color: transparent;
  border: none;
  outline: none;
  color: #fff;
  font-size: 11pt;
  padding-left: 6px;

  &:focus,
  &:active {
    outline: none;
  }
  &::placeholder {
    color: rgba(255, 255, 255, 0.3);
  }
`;

export const IconContainer = styled.span`
  flex: 10%;
  height: 100%;
  display: flex;
  align-items: center;
  padding-left: 8px;
`;

const UserBackground = styled.span`
  padding: 5px;
  border-radius: 50%;
  background-color: ${(props) =>
    props.dark ? 'rgba(168, 168, 168, 0.7)' : 'rgba(255, 255, 255, 0.5)'};
  display: flex;
  align-items: center;
`;

const Icon = styled(PersonIcon)`
  font-size: 27px !important;
  color: ${(props) => (props.dark ? '#4f4f4f' : '#fff')};
`;

export const UserAvatar = ({ dark }) => {
  return (
    <UserBackground dark={dark}>
      <Icon dark={dark ? 1 : 0} />
    </UserBackground>
  );
};

const AuthContainer = styled.span`
  width: 43px;
  height: 43px;
  padding: 2px;
  border-radius: 50%;
  background-color: ${(props) =>
    props.dark ? 'rgba(168, 168, 168, 0.7)' : 'rgba(255, 255, 255, 0.5)'};
  display: flex;
  align-items: center;

  @media screen and (max-width: 600px) {
    width: 40px;
    height: 40px;
  }
`;

const Avatar = styled.img`
  height: 100%;
  border-radius: 50%;
`;

export const AuthAvatar = ({ url }) => {
  return (
    <AuthContainer>
      <Avatar src={url} />
    </AuthContainer>
  );
};
