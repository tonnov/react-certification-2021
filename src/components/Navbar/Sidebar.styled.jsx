import React from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
// import HighlightOffIcon from '@material-ui/icons/HighlightOff';
import CloseIcon from '@material-ui/icons/Close';
import { IconButton } from '@material-ui/core';

const LinkList = styled.ul`
  width: 100%;
  padding: 0;
  margin-top: 8px;
  list-style-type: none;
`;

const LinkItem = styled.li`
  margin: 0;
  padding: 0;
  height: 50px;

  display: flex;
  justify-content: center;
`;

const LinkNav = styled(Link)`
  text-decoration: none;
  /* color: rgb(73, 73, 73); */
  color: inherit;
  font-size: 16px;
  width: 100%;
  height: 100%;
  padding: 18px;
  display: flex;
  align-items: center;

  :hover {
    background-color: #d9d8d8;
  }
`;

const LinkItemNav = ({ item }) => {
  return (
    <LinkItem>
      <LinkNav to={item.route}>{item.name}</LinkNav>
    </LinkItem>
  );
};

const ModalWrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 1000;
  /* display: none; */
  visibility: hidden;
  opacity: 0;
  /* align-items: baseline; */

  &.sidebar-active {
    visibility: visible;
    opacity: 1;
    transition: visibility 0s, opacity 0.2s linear;
  }
`;

const SideBar = styled.nav`
  width: 250px;
  height: 100vh;
  top: 0;
  left: -100%;
  position: absolute;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  padding-top: 10px;
  /* background: #fff; */
  background-color: ${props => props.dark ? '#4d4d4d' : '#fff'};
  color: ${props => props.dark ? '#f2f2f2' : '#4d4d4d'};
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.4);
  z-index: 1050;

  &.sidebar-active {
    transition: 0.5s;
    left: 0;
  }
`;

const FullSideBar = ({ activeClass, updateState, menuItems, dark }) => {
  const handleOnClick = () => {
    updateState();
  };

  return (
    <ModalWrapper className={activeClass}>
      <SideBar className={activeClass} role="menu" aria-label="sidebar" dark={dark}>
        <IconButton onClick={handleOnClick}>
          <CloseIcon style={{ color: 'inherit', fontSize: 30 }} />
        </IconButton>
        <LinkList onClick={handleOnClick}>
          {menuItems &&
            menuItems.map((item) => <LinkItemNav key={item.name} item={item} />)}
        </LinkList>
      </SideBar>
    </ModalWrapper>
  );
};

export default FullSideBar;
