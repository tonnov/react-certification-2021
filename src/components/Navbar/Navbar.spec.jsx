import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Navbar from './Navbar.component';
import AuthProvider from '../../providers/Auth';
import GlobalProvider from '../../providers/Global';

jest.mock('@material-ui/core/Menu', () => () => {
  return (
    <div>
      <input data-testid="mock-component" />
    </div>
  );
});

const component = (
  <GlobalProvider>
    <AuthProvider>
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    </AuthProvider>
  </GlobalProvider>
);

beforeEach(() => render(component));

describe('Navbar Component', () => {
  it('should match snapshot', () => {
    const snap = renderer.create(component);

    expect(snap.toJSON()).toMatchSnapshot();
  });

  it('Should render a menu bar', () => {
    const menuBars = screen.getByRole('button', {
      name: 'open drawer',
    });

    expect(menuBars).toBeInTheDocument();
  });

  it('should render a search input', () => {
    const searchBox = screen.getByRole('textbox', {
      name: 'search',
    });

    expect(searchBox).toBeInTheDocument();
  });

  it('should render a checkbox for theme switch', () => {
    const checkbox = screen.getByRole('checkbox', {
      name: 'Dark mode',
    });

    expect(checkbox).toBeInTheDocument();
  });
});

describe('Navbar User Interaction', () => {
  it('Input should change should not be empty', () => {
    const searchBox = screen.getByRole('textbox', { name: 'search' });
    expect(searchBox.value).not.toBe('');
  });

  it('Input should change query value', () => {
    const searchBox = screen.getByRole('textbox', { name: 'search' });
    const preValue = searchBox.value;
    fireEvent.change(searchBox, { target: { value: `${preValue} Rocks!` } });
    expect(searchBox.value).toBe('wizeline Rocks!');
  });

  it('check if sidebar shows on click', () => {
    const openMenu = screen.getByRole('button', { name: 'open drawer' });
    fireEvent.click(openMenu);
    const sideBar = screen.getByLabelText('sidebar');
    expect(sideBar).toHaveClass('sidebar-active');
  });
});
