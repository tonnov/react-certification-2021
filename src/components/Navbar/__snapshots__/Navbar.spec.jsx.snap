// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component should match snapshot 1`] = `
<nav
  aria-label="NavBar"
  className="sc-bdnxRM liiiJF"
>
  <div>
    <input
      data-testid="mock-component"
    />
  </div>
  <div
    className="sc-gtsrHT klIraI"
  >
    <button
      aria-label="open drawer"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </span>
    </button>
    <div
      className="sc-bqGGPW bCQyud"
      onClick={[Function]}
    >
      <nav
        aria-label="sidebar"
        className="sc-ksluID hUtALt"
        role="menu"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              style={
                Object {
                  "color": "#616161",
                  "fontSize": 30,
                }
              }
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
        </button>
        <span
          className="sc-dIsUp eOLNGD"
        >
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              className="MuiSwitch-root"
            >
              <span
                aria-disabled={false}
                className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiSwitch-switchBase MuiSwitch-colorPrimary"
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={null}
              >
                <span
                  className="MuiIconButton-label"
                >
                  <input
                    checked={false}
                    className="PrivateSwitchBase-input-5 MuiSwitch-input"
                    name="toggleTheme"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  <span
                    className="MuiSwitch-thumb"
                  />
                </span>
              </span>
              <span
                className="MuiSwitch-track"
              />
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Dark mode
            </span>
          </label>
        </span>
        <ul
          className="sc-kEqXSa jUGqmn"
          onClick={[Function]}
        >
          <li
            className="sc-iqAclL ckeQNY"
          >
            <a
              className="sc-crzoAE gBaYqY"
              href="/"
              onClick={[Function]}
            >
              Home
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <div
      className="sc-eCApnc isSkVs"
    >
      <span
        className="sc-gKAaRy iXzfoV"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </span>
      <input
        aria-label="search"
        className="sc-jSFjdj fsuZxZ"
        onChange={[Function]}
        placeholder="Search..."
        value="wizeline"
      />
    </div>
  </div>
  <div
    className="sc-dlnjwi bpbsKY"
  >
    <span
      className="sc-hKFxyN bijRJo"
    >
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-5 MuiSwitch-input"
                name="toggleTheme"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Dark mode
        </span>
      </label>
    </span>
    <button
      aria-label="user"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <span
          className="sc-iCoGMd hGktbA"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root sc-fujyAs cbBOTO"
            dark={0}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</nav>
`;
